### DISCUSSION TITLE
Upgrade guide to 5.0 misleading, causing auth routes to be registered by Nova.

### DISCUSSION THREAD

**Discussion by studnitz** [laravel/nova-issues#6619] [Q&A] [answered]
- Laravel Version: 11.36.1
- Nova Version: 5.0.2
- PHP Version: 8.3.14
- Fortify used on main app: Yes

### Description:

After using the Upgrade guide to upgrade to Nova 5.0, all Fortify auth routes are registered/taken over by Nova and not my app. After running `php artisan route:list` the `login`, `logout` routes are missing and instead are found under `nova/login`, `nova/logout`.

### Detailed steps to reproduce the issue on a fresh Nova installation:

I upgraded my Nova installation by following the steps. One steps says to upgrade `NovaServiceProvider.php` like this:

```diff
/**
 * Register any application services.
 */
public function register(): void 
{
+    parent::register();

    //
}
```

This calls `Nova::fortify()`.

Thing is I already had this in `NovaServiceProvider.php`:

```php
/**
 * Register the Nova routes.
 */
protected function routes(): void
{
    Nova::routes()
        ->withAuthenticationRoutes()
        ->withPasswordResetRoutes()
        ->register();
}
```

where `->register()` does this:

```php
/**
 * Register the Nova routes.
 *
 * @return $this
 */
public function register()
{
    Nova::fortify();

    return $this;
}
```

So, `Nova::fortify()` gets called twice? 

The fix for me was **not** calling `parent::register()` in the `register()` method.

**Comment by crynobone** [comment, marked as answer]
`routes()` method calls `Nova::routes()` while `register()` method called `fortify()` method and therefore `Nova::fortify()`

To solve this you can override the available **undocumented** method:

```php
/**
 * Register the configurations for Laravel Fortify.
 */
protected function fortify(): void
{
    Nova::fortify()
        ->register(routes: true);
}
```

**Reply by crynobone** [reply]
Can you verify that your application does have any of the following service providers?

* `App\Providers\FortifyServiceProvider`
* `App\Providers\JetstreamServiceProvider`

**Reply by studnitz** [reply]
Okay thanks, but maybe this should be also present in the upgrade guide? Because we didn't touch the `NovaServiceProvider` there at all, we just had also Fortify present.

**Reply by studnitz** [reply]
> Can you verify that your application does have any of the following service providers?
> 
> * `App\Providers\FortifyServiceProvider`
> * `App\Providers\JetstreamServiceProvider`

Yes, we have `App\Providers\FortifyServiceProvider`

**Reply by crynobone** [reply]
it should already detect if one of the service provider exists:

![CleanShot 2024-12-18 at 19 12 44](https://github.com/user-attachments/assets/a6b5d16f-f810-46f2-9882-5f5332508def)

but of course, explicitly setting the value would reduce the loading time.

**Reply by studnitz** [reply]
Hmm, `app()->getLoadedProviders()` gives me this:

```
"Illuminate\Events\EventServiceProvider" // vendor/laravel/nova/src/PendingFortifyConfiguration.php:327
"Illuminate\Log\LogServiceProvider" // vendor/laravel/nova/src/PendingFortifyConfiguration.php:327
"Illuminate\Log\Context\ContextServiceProvider" // vendor/laravel/nova/src/PendingFortifyConfiguration.php:327
"Illuminate\Routing\RoutingServiceProvider" // vendor/laravel/nova/src/PendingFortifyConfiguration.php:327
"Illuminate\Auth\AuthServiceProvider" // vendor/laravel/nova/src/PendingFortifyConfiguration.php:327
"Illuminate\Cookie\CookieServiceProvider" // vendor/laravel/nova/src/PendingFortifyConfiguration.php:327
"Illuminate\Database\DatabaseServiceProvider" // vendor/laravel/nova/src/PendingFortifyConfiguration.php:327
"Illuminate\Encryption\EncryptionServiceProvider" // vendor/laravel/nova/src/PendingFortifyConfiguration.php:327
"Illuminate\Filesystem\FilesystemServiceProvider" // vendor/laravel/nova/src/PendingFortifyConfiguration.php:327
"Illuminate\Foundation\Providers\FormRequestServiceProvider" // vendor/laravel/nova/src/PendingFortifyConfiguration.php:327
"Illuminate\Testing\ParallelTestingServiceProvider" // vendor/laravel/nova/src/PendingFortifyConfiguration.php:327
"Illuminate\Foundation\Providers\FoundationServiceProvider" // vendor/laravel/nova/src/PendingFortifyConfiguration.php:327
"Illuminate\Notifications\NotificationServiceProvider" // vendor/laravel/nova/src/PendingFortifyConfiguration.php:327
"Illuminate\Pagination\PaginationServiceProvider" // vendor/laravel/nova/src/PendingFortifyConfiguration.php:327
"Illuminate\Session\SessionServiceProvider" // vendor/laravel/nova/src/PendingFortifyConfiguration.php:327
"Illuminate\View\ViewServiceProvider" // vendor/laravel/nova/src/PendingFortifyConfiguration.php:327
"Clickbar\Magellan\MagellanServiceProvider" // vendor/laravel/nova/src/PendingFortifyConfiguration.php:327
"Clickbar\SapCloudApiClient\SapCloudApiClientServiceProvider" // vendor/laravel/nova/src/PendingFortifyConfiguration.php:327
"Clickbar\Spaceship\SpaceshipServiceProvider" // vendor/laravel/nova/src/PendingFortifyConfiguration.php:327
"Inertia\ServiceProvider" // vendor/laravel/nova/src/PendingFortifyConfiguration.php:327
"Clockwork\Support\Laravel\ClockworkServiceProvider" // vendor/laravel/nova/src/PendingFortifyConfiguration.php:327
"LaravelLang\Config\ServiceProvider" // vendor/laravel/nova/src/PendingFortifyConfiguration.php:327
"LaravelLang\Lang\Plugin" // vendor/laravel/nova/src/PendingFortifyConfiguration.php:327
"LaravelLang\Lang\ServiceProvider" // vendor/laravel/nova/src/PendingFortifyConfiguration.php:327
"LaravelLang\Locales\ServiceProvider" // vendor/laravel/nova/src/PendingFortifyConfiguration.php:327
"LaravelLang\Publisher\ServiceProvider" // vendor/laravel/nova/src/PendingFortifyConfiguration.php:327
"NotificationChannels\WebPush\WebPushServiceProvider" // vendor/laravel/nova/src/PendingFortifyConfiguration.php:327
"Laravel\Fortify\FortifyServiceProvider" // vendor/laravel/nova/src/PendingFortifyConfiguration.php:327
"Laravel\Horizon\HorizonServiceProvider" // vendor/laravel/nova/src/PendingFortifyConfiguration.php:327
"Laravel\Nova\NovaCoreServiceProvider" // vendor/laravel/nova/src/PendingFortifyConfiguration.php:327
"Laravel\Sanctum\SanctumServiceProvider" // vendor/laravel/nova/src/PendingFortifyConfiguration.php:327
"Maatwebsite\Excel\ExcelServiceProvider" // vendor/laravel/nova/src/PendingFortifyConfiguration.php:327
"Carbon\Laravel\ServiceProvider" // vendor/laravel/nova/src/PendingFortifyConfiguration.php:327
"NunoMaduro\Collision\Adapters\Laravel\CollisionServiceProvider" // vendor/laravel/nova/src/PendingFortifyConfiguration.php:327
"Termwind\Laravel\TermwindServiceProvider" // vendor/laravel/nova/src/PendingFortifyConfiguration.php:327
"Pest\Laravel\PestServiceProvider" // vendor/laravel/nova/src/PendingFortifyConfiguration.php:327
"Propaganistas\LaravelPhone\PhoneServiceProvider" // vendor/laravel/nova/src/PendingFortifyConfiguration.php:327
"Rap2hpoutre\FastExcel\Providers\FastExcelServiceProvider" // vendor/laravel/nova/src/PendingFortifyConfiguration.php:327
"Spatie\Activitylog\ActivitylogServiceProvider" // vendor/laravel/nova/src/PendingFortifyConfiguration.php:327
"Spatie\LaravelData\LaravelDataServiceProvider" // vendor/laravel/nova/src/PendingFortifyConfiguration.php:327
"Spatie\MediaLibrary\MediaLibraryServiceProvider" // vendor/laravel/nova/src/PendingFortifyConfiguration.php:327
"Spatie\Navigation\NavigationServiceProvider" // vendor/laravel/nova/src/PendingFortifyConfiguration.php:327
"Spatie\LaravelOptions\OptionsServiceProvider" // vendor/laravel/nova/src/PendingFortifyConfiguration.php:327
"Spatie\LaravelTypeScriptTransformer\TypeScriptTransformerServiceProvider" // vendor/laravel/nova/src/PendingFortifyConfiguration.php:327
"Spatie\StructureDiscoverer\StructureDiscovererServiceProvider" // vendor/laravel/nova/src/PendingFortifyConfiguration.php:327
"Tighten\Ziggy\ZiggyServiceProvider" // vendor/laravel/nova/src/PendingFortifyConfiguration.php:327
"Tpetry\PostgresqlEnhanced\PostgresqlEnhancedServiceProvider" // vendor/laravel/nova/src/PendingFortifyConfiguration.php:327
"App\Providers\AppServiceProvider" // vendor/laravel/nova/src/PendingFortifyConfiguration.php:327
"App\Providers\HorizonServiceProvider" // vendor/laravel/nova/src/PendingFortifyConfiguration.php:327
```

but in `bootstrap/providers.php` I have this:

```php
<?php

return [
    App\Providers\AppServiceProvider::class,
    App\Providers\HorizonServiceProvider::class,
    App\Providers\NovaServiceProvider::class,
    App\Providers\FortifyServiceProvider::class,
];
```

So the check is never true in that case because `NovaServiceProvider` is called before `FortifyServiceProvider`, and that's why it didn't work?

**Reply by crynobone** [reply]
Will work on the routes detection logic so the upgrade will be effortless, thank you for reporting this.

**Reply by Radiergummi** [reply]
adding to this; in a recent release, a `booted` callback was added to the service provider's `register` method that contains `Nova::fortify()->bootstrap($this->app)`.

This peculiar call caused my (Octane) application to resolve my custom `LoginViewResponse` to that included in Nova, for reasons I don't fully understand. It looks like the binding, although scoped, is not flushed at the end of the request, so if I visit our application's own login page *after navigating to Nova*, the container resolves Fortify's request for `Laravel\Fortify\Contracts\LoginViewResponse` to `Laravel\Nova\Auth\Actions\LoginViewResponse` and that fails to render.

I appreciate the effort to modernise Nova's authentication system, but this upgrade has brought a **huge** amount of pain that was absolutely not covered in the upgrade guide. I would kindly ask the Nova team to be more thorough in your pre-release testing in the future.

**Reply by crynobone** [reply]
> It looks like the binding, although scoped, is not flushed at the end of the request

This is a bug with Octane and should be reported to the relevant repo.

![CleanShot 2025-01-10 at 07 28 11](https://github.com/user-attachments/assets/ae3d3909-3f42-4661-866c-5a588c5c5364)

**Reply by Radiergummi** [reply]
It should indeed, but if I update a Laravel application using Octane, Fortify, and Nova 4 to Nova 5, and suddenly dependency resolution issues crop up, that clearly points to Nova as the culprit. Given that we're talking about the compatibility of first-party Laravel libraries here, I think it's reasonable to expect the Laravel team to test the different combinations of packages to ensure there are no incompatibilities or regressions in releases, and prepare a proper change log that lists *all* changes.  
I wouldn't bring this up for an open source package, but we paid for a Nova license, and that changes the situation.

**Reply by crynobone** [reply]
@Radiergummi If you have any issue with integrating Laravel Nova, Octane and Fortify please submit a bug report with reproducing repository in order us to debug the issue. https://github.com/laravel/nova-issues/issues/new?template=1_Bug_report.md

**Reply by PNardman** [reply]
@crynobone 

We are also struggling with the mentioned issues pointed out by @Radiergummi. In addition to that, just following the upgrade guide, seems to have rendered 2FA from fortify basically useless, so users do not get 2FA prompts anymore, when logging in. Due to. Already tried any suggested fixes in this thread however, none of them worked. 

We decided now to downgrade Nova again, because of this.

**Reply by Radiergummi** [reply]
@PNardman in case it helps; we now have the following in our `NovaServiceProvider`, effectively disabling the Nova/Fortify integration completely:

```php
    protected function bootAuthentication(): void
    {
        $this->gate();

        // We're not using the default Fortify authentication, but our own
        // Microsoft Entra ID OAuth integration.
        // Keeping Nova's Fortify integration enabled causes issues with DI
        // container bindings.
        // Nova::fortify()->bootstrap($this->app);
    }

    protected function fortify(): void
    {
        Nova::fortify()->register(routes: false);
    }

    protected function routes(): void
    {
        Nova::routes()
            ->withoutAuthenticationRoutes(false, false)
            ->withoutPasswordResetRoutes(false, false);

        // Instead of using the default authentication routes, we'll register our own
        // routes that use our custom, Socialite-enabled controller.
        // This allows us to defer authentication to Microsoft Azure AD, so employees
        // can log in using their company credentials.
        Nova::router()->group(function (Router $router) {
            $router->get('/login', [EmployeeAuthenticationController::class, 'login'])->name('nova.pages.login');
            $router->get('/login/sso', [EmployeeAuthenticationController::class, 'redirect'])->name('nova.login.sso');
            $router->get('/login/callback', [EmployeeAuthenticationController::class, 'callback'])->name('nova.login');
            $router->post('/logout', [EmployeeAuthenticationController::class, 'logout'])->name('nova.logout');
        });
    }
```

I have pinned Nova to the patch version now, and will probably do manual diffs on Nova releases to catch any changes to this part of the code. It's an unmaintainable mess; you're probably better off on V4.

**Reply by crynobone** [reply]
@PNardman Submit a new bug report with reproducing repository

**Reply by PNardman** [reply]
@crynobone  Will try to. I think the main issue is the upgrade guide and it not properly explaining what the NovaServiceProvider does, so if your application deviates in any way from the default setup, you just have no way of figuring out what you need to do to keep it working.

**Reply by PNardman** [reply]
@Radiergummi Thanks for the input, will try this in a separate way, maybe this would make us able to upgrade, but I agree that it's just a whole mess.

**Reply by crynobone** [reply]
> so if your application deviates in any way from the default setup

This is why reproducing a repository is important if we want to improve the documentation.

**Reply by PNardman** [reply]
@crynobone 

I figured out now why it's overriding the 2FA, I think in recent update it got at least fixed, when you install it from scratch. 

When comparing the upgrade with just installing Nova 5 from scratch you can see that the fortify method is actually added to the NovaServiceProvider like this. 

```
    /**
     * Register the configurations for Laravel Fortify.
     */
    protected function fortify(): void
    {
        Nova::fortify()
            ->features([
                Features::updatePasswords(),
                // Features::emailVerification(),
                // Features::twoFactorAuthentication(['confirm' => true, 'confirmPassword' => true]),
            ])
            ->register();
    }
```

Which makes sure that TwoFactorAuthentication of the own fortify does not get overridden. In general this is not an issue, when you have your own FortifyServiceProvider, however this can lead to issues when also using octane, because of the way the Nova app is bootstrapped and then falsely registers the container services. Which leads to problems in octane. 

I think in general it would help a lot if the documentation could add a bit more explanation in regards to the newly added fortify and bootstrapping, since it seems that the upgrade actually changed a lot under the hood. 

I hope this makes sense.

@Radiergummi maybe this also helps you.
